### ISSUE TITLE
llama-13b finetuning on G5.48x with LORA not working (batch size 1, #trainable parameters ~ 6M)

### ISSUE THREAD

**Issue by vivekmadan2** [facebookresearch/llama-recipes#37]
Hi, I am trying to fine-tune LLama 13B model on g5.48xlarge instance type (8 GPUs, 24GB per GPU)  and running into CUDA OOM issue. However, when I try to check the memory statistics, they don't make sense to me. 

- Loading the model requires less than 30GB GPU memory
- After forward pass takes ~48GB total GPU memory (6GB per GPU)
- Number of trainable params: 6,553,600 (0.05032552357220002% of the total parameters)

Why does the backward pass consume more than 18*8 = 144GB GPU memory for such few parameters?


Setting: torchrun --nnodes 1 --nproc_per_node 8  llama_finetuning.py --use_peft --peft_method lora --model_name /PATH/TO/13B --pure_bf16 --batch_size_training 1 --micro_batch_size 1 --enable_fsdp


 File "llama_finetuning.py", line 237, in <module>                                                                                                                                                           [214/1853]
    fire.Fire(main)                                                                                                                                                                                                     
  File "/home/ubuntu/.local/lib/python3.8/site-packages/fire/core.py", line 141, in Fire                                                                                                                                
    component_trace = _Fire(component, args, parsed_flag_args, context, name)                                                                                                                                           
  File "/home/ubuntu/.local/lib/python3.8/site-packages/fire/core.py", line 475, in _Fire                                                                                                                               
    component, remaining_args = _CallAndUpdateTrace(                                                                                                                                                                    
  File "/home/ubuntu/.local/lib/python3.8/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace                                                                                                                 
    component = fn(*varargs, **kwargs)                                                                                                                                                                                  
  File "llama_finetuning.py", line 220, in main                                                                                                                                                                         
    results = train(                                                                                                                                                                                                    
  File "/home/ubuntu/llama-recipes/utils/train_utils.py", line 110, in train                                                                                          
    loss.backward()                                                                                                                                                                                                     
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/_tensor.py", line 487, in backward                                                                                                                        
    torch.autograd.backward(                                                                                                                                                                                            
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/autograd/__init__.py", line 200, in backward                                                                                                              
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass                                                                                                                      
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 643, in _pre_backward_hook                                                                                      
    _prefetch_handles(state, _handles_key)                                                                                                                                                                              
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 1007, in _prefetch_handles                                                                                      
    _unshard(                                                                                                                                                                                                           
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 329, in _unshard                                                                                                
    handle.unshard()                                                                                                                                                                                                    
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/fsdp/flat_param.py", line 918, in unshard                                                                                                     
    unsharded_flat_param = self._alloc_padded_unsharded_flat_param()                                                                                                                                                    
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/fsdp/flat_param.py", line 944, in _alloc_padded_unsharded_flat_param
    _alloc_storage(unsharded_flat_param, flat_param._padded_unsharded_size)  # type: ignore[attr-defined]                                                                                                               
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/fsdp/_utils.py", line 79, in _alloc_storage
    tensor._typed_storage()._resize_(size.numel())
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/storage.py", line 787, in _resize_

**Comment by HamidShojanazeri** [comment]
@vivekmadan2 are you using PT nightlies? reason is fsdp was not able to free the parameters that their `require_grad=False`, when using PEFT/LORA the model is mostly frozen, the fix was landed on nightlies few weeks ago. Can you pls give it a try if not on nightlies yet.

**Comment by vivekmadan2** [comment]
@HamidShojanazeri I tried installing torch nightly and fine-tuning again but the error persists.

Installing torch nightly: pip3 install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu118

**Comment by vivekmadan2** [comment]
Verified that using the nightly version 

Here is the output of 
python3 -m "torch.utils.collect_env"


PyTorch version: 2.1.0.dev20230724+cu118              
Is debug build: False           
CUDA used to build PyTorch: 11.8                  
ROCM used to build PyTorch: N/A                       
                                                      
OS: Ubuntu 20.04.6 LTS (x86_64)                   
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect                                                                                                                                                                                        
CMake version: version 3.26.4                                                                               
Libc version: glibc-2.31                              
                                                                                                                                                                                                                        
Python version: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-1037-aws-x86_64-with-glibc2.31
Is CUDA available: True      
CUDA runtime version: 11.8.89   
CUDA_MODULE_LOADING set to: LAZY
------

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.1
[pip3] pytorch-triton==2.1.0+9e3e10c5ed
[pip3] torch==2.1.0.dev20230724+cu118
[conda] numpy                     1.25.1                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+9e3e10c5ed          pypi_0    pypi
[conda] torch                     2.1.0.dev20230724+cu118          pypi_0    pypi

**Comment by HamidShojanazeri** [comment]
@vivekmadan2 this PR should improve the memory situation, can pls give it another try https://github.com/facebookresearch/llama-recipes/pull/41

**Comment by vivekmadan2** [comment]
@HamidShojanazeri Issue persists for me even with the updated code.

**Comment by HamidShojanazeri** [comment]
@vivekmadan2 I am running on A100, dont have my A10 setup yet but the memory number should be extensible to A10.

running this command on 8xA100 have this mem stats:

```
torchrun --nnodes 1 --nproc_per_node 8  llama_finetuning.py  --model_name PATH/TO/13B   --use_peft --peft_method lora  --enable_fsdp --pure_bf16 True  --output_dir 13b_fdsp_lora
```

```
Max CUDA memory allocated was 13 GB
Max CUDA memory reserved was 17 GB
Peak active CUDA memory was 13 GB
```

**Comment by vivekmadan2** [comment]
How would you recommend me progressing here because the issues are still persisting with the configurations provided above?

**Comment by HamidShojanazeri** [comment]
@vivekmadan2 from the other issue you opened #45, it seems training is running and now OOM is on model saving part. One other thing I can suggest is to either lower the batch size or alternatively using [multi-node](https://github.com/facebookresearch/llama-recipes/tree/main#multi-gpu-multi-node), I believe g5.48xlarge support EFA but if not please make sure to remove EFA related parts.

**Comment by vivekmadan2** [comment]
what is EFA? Sorry, I am not sure what you mean by EFA related parts?

Training is running without LORA. I am still not able to train with LORA+FSDP. I would like that option especially going forward when fine-tuning 70B model. 

Trying on smaller batch size to see if that resolves the issue without LORA.

**Comment by HamidShojanazeri** [comment]
@vivekmadan2 can you pls add this env variable to the llama-finetuning.py, it should help with memory fragmentation that might contribute to this OOM as well, I will add it as a default to the script in PR.

```
import os
os.env['PYTORCH_CUDA_ALLOC_CONF']='expandable_segments:True'
```

**Comment by HamidShojanazeri** [comment]
> what is EFA? Sorry, I am not sure what you mean by EFA related parts? 

EFA is an fast network interface for AWS EC2 instances.

>I am still not able to train with LORA+FSDP.

Not clear about it, can you please share more details the command you are running?

**Comment by HamidShojanazeri** [comment]
You can also refer to this issue, seems LORA+FSDP could have been run on 8xV100, https://github.com/facebookresearch/llama-recipes/issues/36

**Comment by vivekmadan2** [comment]
I got LORA+FSDP working on G5.48x by properly installing and configuring torch-nightly. 

@HamidShojanazeri May be update requirements.txt so that it doesn't become an issue for anyone else?

**Comment by HamidShojanazeri** [comment]
@vivekmadan2  thanks for the updates, it requirements we defaulted to stable but added note [here](https://github.com/facebookresearch/llama-recipes/tree/main#multiple-gpus-one-node)
